### PR TITLE
Equal odds postprocessing workflow and blog

### PR DESCRIPTION
**Pull request by ZanMervic**
Fifth of fairness workflows and blogs.

### CODE REVIEW

File: content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
++++
+author = "Žan Mervič"
+date = "2023-08-30"
+draft = false
+title = "Orange Fairness - Equal Odds Postprocessing"
+type = "blog"
+thumbImage = "/blog_img/2023/2023-08-30-fairness-equal-odds-postprocessing.png"
+frontPageImage = "/blog_img/2023/2023-08-30-fairness-equal-odds-postprocessing.png"
+blog = ["fairness", "equal odds postprocessing"]
+shortExcerpt = "In this blog, we delve into the Equal Odds Postprocessing widget, a tool designed to enhance fairness in machine learning models. We break down how the algorithm works by modifying predictions to meet Equalized Odds criteria. Using a real-world example with the German credit dataset, we demonstrate its efficacy in improving fairness metrics while marginally affecting accuracy."
```
Comment (markotoplak) [inline content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md:10]: short < long

File: content/workflows/fairness-equal-odds-postprocessing.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+images =  ["/workflows/images/fairness-equal-odds-postprocessing.png"]
+type = "workflows"
+blog_link =  "/blog/2023/2023-08-30-fairness-equal-odds-postprocessing/"
+video = ""
+download = "fairness-equal-odds-postprocessing.ows"
+workflows = ["Fairness"]
+weight = 1004
++++
+
+Another way to mitigate bias besides the ones we have shown before is to use a postprocessing algorithm on the model's predictions. This workflow illustrates using the Equal Odds widget as a post-processor for the Logistic Regression model. To use the post-processor, we need to connect any model to the Equalized Odds Postprocessing widget along with any needed pre-processors. Doing so ensures our model's predictions get post-processed before we evaluate them. We then connect the Equalized Odds Postprocessing widget to the Test and Score widget and visualize the results using a Mosaic Display.
```
Comment (markotoplak) [inline content/workflows/fairness-equal-odds-postprocessing.md:12]: "besides the ones we have shown before" - workflows should work by themselvers, do not reference others

Comment (markotoplak) [reply content/workflows/fairness-equal-odds-postprocessing.md:12]: For me, everything after "Doing so" is fairly obvious, non-specific to this use-case, and should thus be removed.

File: content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
++++
+
+
+In the [previous blog post]({{< ref "2023-08-28-fairness-adversarial-debiasing.md" >}}), we discussed the Adversarial Debiasing model, a bias-aware model. This blog post will discuss the Equal Odds Postprocessing widget, a bias-aware post-processor, which can be used with any model to mitigate bias in its predictions.
+
+### Equal Odds Postprocessing:
+
+The [Equal Odds Postprocessing](https://arxiv.org/abs/1610.02413) widget is a post-processing type of fairness mitigation algorithm designed to ensure fairness in supervised learning. It modifies the predictions of any given classifier to meet certain fairness criteria, specifically focusing on "Equalized Odds" or more relaxed criteria like Equal Opportunity. Because it is a post-processing algorithm, it does not require any support from the model (like reweighing, which requires the model to support weights) or changes to the model (like adversarial debiasing, which requires the model to be a neural network). This makes it a versatile algorithm that can be used with any model.
```
Comment (markotoplak) [inline content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md:19]: designed to ensure fairness in supervised learning -> "for supervised learning" (the reader knows from the rest of the sentence that this is about fairness.

any support from the model - strange and unclear phrasing, I'd do this:
"it does not require any support from the model (like reweighing, which requires the model to support weights) or changes to the model (like adversarial debiasing, which requires the model to be a neural network). This makes it a versatile algorithm that can be used with any model." -> "it works with general supervised models."

File: content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+The widget Equalized Odds Postprocessing represents a novelty in the Orange environment, as until now, there has been no widget with similar functionality. Its operation is similar to other Orange widgets representing models, except it also expects one of the Orange models as an input. It operates by first learning from the predictions of the model's training data. Then, it applies the Equalized Odds Postprocessing algorithm from the [AIF360](https://aif360.res.ibm.com/) library to the predictions the selected model makes on the test data. This ensures that the adjusted predictions conform to the fairness definition of equalized odds.
```
Comment (markotoplak) [inline content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md:21]: represents a novelty - too ornamental, also, if something is novel, then "as until now, there has been no widget with similar functionality" is a tautology and thus needless.

"Its operation is similar to other Orange widgets representing models, except it also expects one of the Orange models as an input." -> "It needs a model as an input." (also, does it need a learner or model?)

"It operates by first learning from the predictions of the model's training data." - I have not idea what you wanted to say with this

"This ensures that the adjusted predictions conform to the fairness definition of equalized odds." -> This is already said with the algorithm name, why repeat it here?

File: content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+In more detail, the Equalized Odds Postprocessing algorithm works by solving a linear program with some constraints and the following objective function:
```
Comment (markotoplak) [inline content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md:23]: "In more detail" -> You lose no detail by removing it

File: content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+`c = [fpr_0 - tpr_0, tnr_0 - fnr_0, fpr_1 - tpr_1, tnr_1 - fnr_1]`
+
+Where `fpr`, `tpr`, `tnr`, and `fnr` are the false positive, true positive, true negative, and false negative rates for privileged (`0`) and unprivileged (`1`) groups.
+
+The algorithm then finds the optimal solution to the linear program, which results in a set of probabilities with which to flip the model's predictions to equalize the odds of being correctly or incorrectly classified for both privileged and unprivileged groups:
+
+- `sp2p`: Probability of flipping a label from positive to negative for the privileged group.
```
Comment (markotoplak) [inline content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md:31]: Probability of flipping a label is repeating. Reformulate to avoid such repetitions.

File: content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+- `sn2p`: Probability of flipping a label from negative to positive for the privileged group.
+- `op2p`: Probability of flipping a label from positive to negative for the unprivileged group.
+- `on2p`: Probability of flipping a label from negative to positive for the unprivileged group.
+
+
+## Orange use case
```
Comment (markotoplak) [inline content/blog/2023/2023-08-30-fairness-equal-odds-postprocessing.md:37]: I did not read further, but try to rewrite. :)
